src/Data/RFC5322.hs

```diff
@@ -124,8 +131,82 @@ special = satisfy isSpecial
 atext :: Parser Word8
 atext = satisfy isAtext
 
+
+-- §3.4 Address Specification
+data Mailbox =
+    Mailbox (Maybe T.Text {- display name -})
+            B.ByteString {- addr spec -}
+    deriving (Show,Eq)
+
+mailbox :: Parser Mailbox
+mailbox = Mailbox <$> optionalDisplayName <*> angleAddr
+
+optionalDisplayName :: Parser (Maybe T.Text)
+optionalDisplayName = option Nothing (Just <$> (optionalCFWS *> displayName <* optionalCFWS))
+
+displayName :: Parser T.Text
+displayName = decodeLenient <$> phrase
+
+word :: Parser B.ByteString
+word = atom <|> quotedString
+
+phrase :: Parser B.ByteString
+phrase = foldMany1 word
+
+angleAddr :: Parser B.ByteString
+angleAddr = optionalCFWS *>
+  char8 '<' *> addressSpec <* char8 '>'
+  <* optionalCFWS
+
+atom :: Parser B.ByteString
+atom = optionalCFWS *> foldMany1 (B.singleton <$> atext) <* optionalCFWS
+
+dotAtomText :: Parser B.ByteString
+dotAtomText = foldMany1 (B.singleton <$> atext) <<>> foldMany ((B.singleton <$> char8 '.') <<>> takeWhile1 isAtext)
+
+dotAtom :: Parser B.ByteString
+dotAtom = optionalCFWS *> dotAtomText <* optionalCFWS
+
+addressSpec :: Parser B.ByteString
+addressSpec = localPart <<>> (B.singleton <$> char8 '@') <<>> domain
+
+localPart :: Parser B.ByteString
+localPart = dotAtom <|> quotedString
+
+-- | Printable US-ASCII excl "[", "]", or "\"
+isDtext :: Word8 -> Bool
+isDtext c = (c >= 33 && c <= 90) || (c >= 94 && c <= 126)
+
+dText :: Parser Word8
+dText = satisfy isDtext
+
+domainLiteral :: Parser B.ByteString
+domainLiteral =
+  optionalCFWS *> char8 '['
+  *> foldMany (optionalFWS <<>> (B.singleton <$> dText) <<>> optionalFWS)
+  <* char8 ']' <* optionalFWS
```

Comment:
Here we're discarding the indicator that this is a domain literal.  Because the acceptable domain literal values are context-sensitive, it's possible that there is no way to distinguish.  So I think we should have a proper tagged union.